rpc/src/module/ipc.rs

```diff
@@ -0,0 +1,883 @@
+            .map_err(|e| RPCError::custom_with_error(RPCError::TransactionFailedToResolve, e))?;
+        // Due to the existence of system cell cache, some system scripts will not be obtained from Resource. There is
+        // currently no way to disable the system cell cache, so manually override them here.
+        for e in tx.resolved_cell_deps.iter_mut() {
+            if let Some(CellStatus::Live(meta)) = dl.cell.get(&e.out_point) {
+                *e = meta.clone()
+            }
+        }
+        let snapshot = self.shared.cloned_snapshot();
+        let pipesctx = PipesCtx::default();
+        let script_verifier = TransactionScriptsVerifier::<_, _, Machine>::new_with_generator(
+            Arc::new(tx),
+            dl,
+            snapshot.cloned_consensus(),
+            Arc::new(TxVerifyEnv::new_submit(snapshot.tip_header())),
+            overload_ckb_syscalls,
+            pipesctx.clone(),
+        );
+        let script_group = script_verifier
+            .find_script_group(
+                match env.script_group_type {
+                    ScriptGroupType::Lock => ckb_script::ScriptGroupType::Lock,
+                    ScriptGroupType::Type => ckb_script::ScriptGroupType::Type,
+                },
+                &env.script_hash.pack(),
+            )
+            .ok_or_else(|| {
+                RPCError::custom_with_error(
+                    RPCError::IPC,
+                    ckb_script::ScriptError::ScriptNotFound(env.script_hash.pack()),
+                )
+            })?
+            .clone();
+
+        // In any case, we will close all pipes after a certain period of time.
+        // This ensures that the function always completes if the ipc script is not written as expected.
+        let pipesfin = pipesctx.clone();
+        let mut signal_timeout = tokio::sync::watch::channel(0);
+        let limit_time = self.limit_time;
+        self.shared.async_handle().spawn({
+            async move {
+                let _ = tokio::time::timeout(
+                    tokio::time::Duration::from_secs(limit_time as u64),
+                    signal_timeout.1.changed(),
+                )
+                .await;
+                pipesfin.close();
+            }
+        });
+
+        let pipesfin = pipesctx.clone();
+        let limit_cycles = self.limit_cycles;
+        let upper_cycles = 1000000;
+        self.shared.async_handle().spawn({
+            async move {
+                // Ignore its return value as it is unimportant.
+                let _ = script_verifier
+                    .create_scheduler(&script_group)
+                    .map(|mut scheduler| {
+                        // Make the vm can always be stopped.
+                        debug_assert_eq!(scheduler.consumed_cycles(), 0);
+                        let tic = std::time::SystemTime::now();
+                        loop {
+                            let last_cycles = limit_cycles as u64 - scheduler.consumed_cycles();
```

Comment:
do we have a risk of overflow here?

Reply:
We are better off using the `saturating_sub` method